### PR TITLE
RET-3397. Implement dynamic links for all templates

### PR DESCRIPTION
**Pull request by Itay-Kirsh**
https://tools.hmcts.net/jira/browse/RET-3397

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

**Review by brendanmchugh — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseService.java [hunk 1]
```diff
@@ -137,10 +142,31 @@ public void saveRelatedDocumentsToDocumentCollection(CaseData caseData) {
...
+        Map<String, String> personalisation = new ConcurrentHashMap<>();
+        personalisation.put("case_number", caseData.getEthosCaseReference());
+        personalisation.put(CLAIMANT, caseData.getClaimant());
+        personalisation.put("list_of_respondents", getRespondentNames(caseData));
+        personalisation.put(DATE, ReferralHelper.getNearestHearingToReferral(caseData, "Not set"));
+        personalisation.put(LINK_TO_EXUI, emailService.getExuiCaseLink(caseDetails.getCaseId()));
+        return personalisation;
+    }
+
+    // todo probably should be replaced with Helper.getRespondentNames
```
Comment (JackReevies) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseService.java:166]: Still relevant?

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseService.java [hunk 2]
```diff
@@ -137,10 +142,31 @@ public void saveRelatedDocumentsToDocumentCollection(CaseData caseData) {
...
+        emailService.sendEmail(
+            et3EmailTribunalTemplateId,
             caseDetails.getCaseData().getTribunalCorrespondenceEmail(),
             buildPersonalisation(caseDetails)
         );
     }
 
+    // todo group all personalisation methods and NotificationServiceConstants to EmailService or to a
+    // PersonalisationService. Might even be able to join into one single method that prepares all the data for all
+    // emails (unless some of it is expensive / bespoke / overrides in which case send a map parameter of overrides)
```
Comment (JackReevies) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseService.java:154]: Still relevant? I'm not sure about the idea of moving all personalisations into a single class. Sounds like the formation of a godclass that requires knowledge of too many things (like TornadoService). the personalisation objects are specific to each email template which is specific to the business needs of that service (ie ET3 form), it should remain up to ET3 related classes to know what data it needs for the template

**Review by JackReevies — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseService.java (hunk 1 above)
Comment (Itay-Kirsh) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseService.java:166]: Yeah, just not relevant to this ticket, it's quite a big thing the personalisation refactor that should happen here

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseService.java (hunk 2 above)
Comment (Itay-Kirsh) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/service/Et3ResponseService.java:154]: All personalisations have a core set of keys and very few have anything that's unique to the specific email being sent. We can make a generic method that adds those core properties and accepts any more unique ones as a parameter. it's just a thought though and if you disagree, I don't feel too strongly about it, we can remove the todo.

**Review by JackReevies — APPROVED** [review]

**Pull request merged**